### ISSUE TITLE
os.symlink: FileExistsError shows wrong message

### ISSUE THREAD

**Issue by d943741b-dd32-4a1a-b522-8ba0d540c2df** [python/cpython#73843]
BPO | [29657](https://bugs.python.org/issue29657)
--- | :---
Nosy | @larryhastings, @serhiy-storchaka, @eryksun, @sayanchowdhury, @zhangyangyu, @andrewnester, @Miserlou, @Jongy, @kamilturek
Files | <li>[x.py](https://bugs.python.org/file46671/x.py "Uploaded as text/plain at 2017-02-26.08:52:23 by wrohdewald")</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2017-02-26.08:52:24.286>
labels = ['interpreter-core', 'type-bug', '3.8', '3.9', '3.10', 'extension-modules']
title = 'os.symlink: FileExistsError shows wrong message'
updated_at = <Date 2021-05-01.14:08:17.914>
user = 'https://bugs.python.org/wrohdewald'
```

bugs.python.org fields:
```python
activity = <Date 2021-05-01.14:08:17.914>
actor = 'Yonatan Goldschmidt'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Extension Modules', 'Interpreter Core']
creation = <Date 2017-02-26.08:52:24.286>
creator = 'wrohdewald'
dependencies = []
files = ['46671']
hgrepos = []
issue_num = 29657
keywords = []
message_count = 12.0
messages = ['288591', '288592', '288906', '288936', '288947', '288963', '288971', '289064', '296270', '296273', '330827', '392598']
nosy_count = 10.0
nosy_names = ['larry', 'wrohdewald', 'serhiy.storchaka', 'eryksun', 'sayanchowdhury', 'xiang.zhang', 'andrewnester', 'miserlou', 'Yonatan Goldschmidt', 'kamilturek']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue29657'
versions = ['Python 3.8', 'Python 3.9', 'Python 3.10']
```

</p></details>

**Comment by d943741b-dd32-4a1a-b522-8ba0d540c2df** [comment]
execute the attached script. It should return

FileExistsError: [Errno 17] File exists: 'a_link' -\> 'a'

but it returns

FileExistsError: [Errno 17] File exists: 'a' -\> 'a_link'

**Comment by zhangyangyu** [comment]
I concur the current message is misleading.

OSError makes the string "[file1](https://bugs.python.org/file1) -> [file2](https://bugs.python.org/file2)". This also affects other methods calling `path_error2()` such as os.link().

**Comment by 85518756-5bea-4efa-9ce9-5daf552d31b1** [comment]
I've been investigating this issue and did not come up with some easy solution.

So the problem is:
os_symlink uses `path_error2` to throw exception.
the order of file arguments now is src then dest. For provided example src is `a` and dest is `sym_link`. As a result `src` -> `dest` is generated here https://github.com/python/cpython/blob/master/Objects/exceptions.c#L1059

If we change order of arguments passed to `path_error2`, error message will be generated properly but OSError.filename will be incorrect (a_link instead of a) and following test will fail for `link`/`symlink`

https://github.com/python/cpython/blob/master/Lib/test/test_os.py#L2901

Not sure if it's OK or not, so it definitely needs some input from Python core developers.

**Comment by serhiy-storchaka** [comment]
The current error message looks good to me. What is wrong with it?

**Comment by eryksun** [comment]
A symbolic link is typically represented the other way around, from the point of view of the link pointing at the target. However, Python conceptualizes linking as something like a copy or rename operation, with source and destination filenames, and the current error message represents this point of view.

**Comment by serhiy-storchaka** [comment]
This is not about how a symbolic link is represented. This is about how the operation of creating a symbolic link is represented. The first filename is the first argument of os.symlink(), the second filename is the second argument. Try to run

```py
    os.symlink('a', 'a_link')
    os.symlink('b', 'a_link')
```


You should get an error:

    FileExistsError: [Errno 17] File exists: 'b' -> 'a_link'

**Comment by eryksun** [comment]
To me the error message is in the model of a source -\> destination operation, in which the arrow indicates the operation's information flow (e.g. of the target path or inode number) from the source file to the destination file. I've never viewed it superficially as just an ordering of parameter1 -\> parameter2.

**Comment by zhangyangyu** [comment]
I agree with Eryk. The error message is misleading. When I see it, I take it as source -\> destination. I think we should make the error message clearer, or document it in the OSError documentation.

**Comment by serhiy-storchaka** [comment]
See bpo-20517 for the discussion about current implementation. Are there any ideas about clearer error messages?

Added Larry as the author of the original idea and implementation.

**Comment by larryhastings** [comment]
"Special cases aren't special enough to break the rules."  I want the error message to mirror the API, which it currently does.  If we swapped them, the error message would now contradict the API.  So no, I don't support swapping "src" and "dst" in the error message only when the error pertains to os.symlink.

If literally every time the two-filename version of OSError is used inside Python, the two filenames are "src" and "dst", then we could consider making it slightly more explicit, e.g.

    FileExistsError: [Errno 17] File exists: src='a', dst='a_link'

I think I'd want the source code to reflect this (e.g. thinking about "src" and "dst" rather than "filename" and "filename2").

Would OP et al consider this change to the error message an improvement, or is it not interesting?

**Comment by pevogam** [comment]
Same thing here, it took me 10-15 mins to figure out what is happening after manually removing both paths on sides and experimenting.

**Comment by JeanLescutMuller** [comment]
Same here. Could someone agree on changing this very confusing error message ?

**Comment by ncoghlan** [comment]
The `FileNotFoundError` raised when the directory the symlink is to be placed in doesn't actually exist is similarly confusing.

```
FileNotFoundError: [Errno 2] No such file or directory: '/some_dir/target' -> '/some_missing_dir/link'
```

Given that symlink creation doesn't care whether or not the link target exists (it only cares about the symlink location itself), perhaps a reasonable resolution would be to use the *single* filename version of the OS error reporting APIs in `os_symlink_impl`?

That would make the error reporting look like:

```
FileExistsError: [Errno 17] File exists: 'a_link'
```
```
FileNotFoundError: [Errno 2] No such file or directory: '/some_missing_dir/link'
```

And the filename info on the `FileNotFoundError` instance would still point to the correct location (the link rather than the link's target)

**Comment by BusiestBeaver** [comment]
Same issue here. Another hour wasted due to misleading error message. This really needs a fix.
